Implement many new functions and some minor cleanups

olsonse opened the pull request:
I've tested and used these patches for the last 6 or so years based on the earlier code.google.com version.  I've now reapplied these patches to the new baseline at github with very few conflicts.  

These patches also include two bug fixes for:
- Task.is_done:
  This patch causes Task.is_done to return the value that is written by DAQmxIsTaskDone, rather than the error code that was returned (which is 0 on success, regardless of whether the task is complete or not).
- CounterInputTask.get_duplicate_count_prevention:
  This patch similarly fixes the return value to be the value written by DAQmxGetCIDupCountPrevent, rather than the error code that was returned.

These patches add interfaces for:
- Device.get_analog_output_sample_clock_supported (DAQmxGetDevAOSampClkSupported)
- System.connect_terminals (DAQmxConnectTerms)
- System.disconnect_terminals (DAQmxDIsconnectTerms)
- System.tristate_terminal (DAQmxTristateOutputTerm)
- Task.get_sample_timing_type (GetSampTimingType)
- Task.set_sample_timing_type (SetSampTimingType)
- Task.reset_sample_timing_type (ResetSampTimingType)
- Task.get_sample_mode (GetSampQuantSampMode)
- Task.set_sample_mode (SetSampQuantSampMode)
- Task.reset_sample_mode  (ResetSampQuantSampMode)
- Task.get_samples_per_channel (GetSampQuantSampPerChan)
- Task.get_samples_per_channel (SetSampQuantSampPerChan)
- Task.reset_samples_per_channel (ResetSampQuantSampPerChan)
- Various routines for Getting/Setting physical channel attributes

Other changes in these patches:
- Task.clear
  Make it called more like the other various DAQmx functions.
- Adds a few more header files for more versions of DAQmx lib

olsonse committed:
small typo fixes

olsonse committed:
change treatment of bools to be more correct

olsonse committed:
implement ResetDevice, ConnectTerms, DisconnectTerms, TristateOutputTerm

olsonse committed:
Add nidaqmx version 8.6 header module

olsonse committed:
change ClearTask to be called like all other nidaqmx calls--easier for simulated hardware

olsonse committed:
add physical channel attribute accessors/setters

olsonse committed:
Add function to test whether AO channels of a device can be hardware-timed

olsonse committed:
add lower level sample timing control/access

olsonse committed:
postpone initialization of some Task class variables that depend on nidaqmx_h being loaded

Postponing until the first Task instantiation allows for nidaqmx_h to be loaded
later--this eases the implementation of a simulated low level library.

olsonse committed:
ignore files for git

olsonse committed:
add generated nidaqmx 8.7 header

Commit:
removed small bug in test_ContGen_IntClk.py

Commit:
fixed bug in .read() method. (samples_per_channel=-1 was not handled appropriately)

Commit:
added digital trigger filters to libnidaqmx.py

pearu commented on nidaqmx/libnidaqmx.py:
```diff
@@ -780,6 +818,45 @@ def __init__(self, name = ""):
+        nidaqmx_h file to be loaded later.  This eases the implementation of a
+        simulated low-level library.
+        """
+        if 'sample_mode_map' in dir(Task):
```
I suggest using `if hasattr(Task, 'sample_mode_map')`

pearu commented on nidaqmx/libnidaqmx.py:
```diff
@@ -780,6 +818,45 @@ def __init__(self, name = ""):
+        """
+        if 'sample_mode_map' in dir(Task):
+          return
+        Task.sample_mode_map = dict (finite = DAQmx.Val_FiniteSamps,
```
Could loading the mappings be implemented without introducing the static method load_maps? (to me load_maps approach looks hackish).

For instance, the attributes sample_mode_map, edge_map, sample_timing_type_map as well as their computed reverses, could all be simply class attributes.
The late loading of nidaqmx_h does not seem to be implemented anyway. If that is the plan, then the mappings could be introduced as instance properties.

olsonse replied:
If a proper DAQmx library is not loaded, then nidaqmx_h is not loaded and neither do the DAQmx constants exist, which is why we cannot simply create these variables at class definition time--otherwise `import nidaqmx` would have failed outright (when DAQmx is not actually available).

I have another package ([Arbwave](https://github.com/afrl-quantum/arbwave)) that must implement simulated versions of the various drivers it talks to (particularly helpful when simulating some experimental configuration when not on a machine with either the proper hardware--or even Windows).

I needed some way of delaying the instantiation of these class variables until after it was very likely that either the real library _or_ the simulated library had been loaded.  It seemed very likely that any loading of real libraries or injecting of simulated ones would occur before any tasks were attempted to be created.

I load the simulated library by importing nidaqmx, then replacing the actual NI libs with:
  `nidaqmx.libnidaqmx.libnidaqmx = sim.NiDAQmx()`
This sim.NiDAQmx class claims to be v8.0 and implements dummy versions of the various DAQmx* library functions that are used in pylibnidaqmx.  After this injection, I reload the nidaqmx_h file and the v8.0 header is picked up.

I wrote this a long time ago, but looking at this, loadmaps probably should have minimally been implemented as a classmethod and not a staticmethod.  It should probably also have a leading '_' to help it be indicated as a privately used tool.

With respect to these being instance properties rather than class properties, I think I might agree, but mostly from the point of being able to de-inject the simulated library and reloading the proper library.

Thoughts?

pearu reviewed (changes requested):
Thanks for submitting this PR! It looks very good. I have made some minor comments regarding the load_maps method.

olsonse commented:
This patch set has been updated, following @pearu's suggestion that the various items in load_maps be moved to instance variables.  As such, _load_maps was removed and this work is now done in the Task constructor.

olsonse committed:
Fix another .read() method when samples_per_channel == DAQmx.Val_Auto

This patch also changes the fixes in commit 96573565544b to use
DAQmx.Val_Auto, so as to better read like the c-interface.

olsonse commented:
Just pushed into this PR a few cherry-picks from @fgoettfert that are definitely applicable to a pre-py3 branch.
